GetSchwifty/Code Helper/Completer.swift

```diff
@@ -0,0 +1,326 @@
+//
+//  Completer.swift
+//  GetSchwifty
+//
+//  Created by Lennart Wisbar on 03.05.17.
+//  Copyright © 2017 Benjamin Herzog. All rights reserved.
+//
+
+import Foundation
+
+struct Completer {
+    static func completedText(for input: String, in text: String, range: NSRange) -> (newText: String, newRange: NSRange) {
+        guard
+            let selection = text.stringRange(from: range),
+            let scenario = Completer.scenario(for: input, in: text, range: range)
+            else { return (text, range) }
+        
+        let line = text.lineRange(for: selection.lowerBound..<selection.lowerBound)
+        let indentation = text.indentationLevel(of: line)
+        
+        var (insertion, cursorOffset) = completedInput(for: input, scenario: scenario, indentation: indentation)
+        
+        var newText = text.replacingCharacters(in: selection, with: insertion)
+        
+        // Change the indentation of the current line if needed
+        // TODO: Refactor indentation correction into its own method?
+        if scenario == .colonAfterCaseOrDefault,
+            let switchIndentation = text.indentationLevelOfLast("switch", before: selection.lowerBound) {
+            let newCursor = newText.index(selection.lowerBound, offsetBy: cursorOffset)
+            let newLine = newText.lineRange(for: newCursor..<newCursor)
+            newText = newText.settingIndentationLevel(of: newLine, to: switchIndentation)
+            cursorOffset += (switchIndentation - indentation)
+        }
+        
+        let newLocation = range.location + cursorOffset
+        let newRange = NSMakeRange(newLocation, 0)
+        
+        return (newText: newText, newRange: newRange)
+    }
+}
+
+private extension Completer {
+    enum Scenario {
+        case normal
+        case newLine
+        case newLineAfterCurlyBrace
+        case newLineBetweenCurlyBraces
+        case newLineAfterCurlyBraceAlreadyClosed
+        case newLineAfterCurlyBraceAfterSwitch
+        case newLineBetweenCurlyBracesAfterSwitch
+        case newLineAfterCurlyBraceAlreadyClosedAfterSwitch
+        case newLineAfterColonAfterCaseOrDefault
+        case colonAfterCaseOrDefault
+        case openRoundBracket
+        case openSquareBracket
+        case closedRoundBracketBeforeClosedBracket
+        case closedSquareBracketBeforeClosedBracket
+        case closedCurlyBraceBeforeClosedCurlyBrace
+        case quotationMark
+        case quotationMarkBeforeQuotationMark
+    }
+    
+    static func completedInput(for input: String, scenario: Scenario, indentation: Int) -> (String, cursorOffset: Int) {
+        var insertion = input
+        var cursorOffset = input.utf16.count
+        
+        switch scenario {
+        case .normal:
+            insertion = input
+            cursorOffset = insertion.utf16.count
```

Comment:
If you would have said to me that we are discussing about emojis in 2017 5 years ago I would had laughed about you :D